Change memory format promotion rules of point wise operators.

VitalyFedyunin opened the pull request:
Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37968 Change memory format promotion rules of point wise operators.**

Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

VitalyFedyunin committed:
Change memory format promotion rules of point wise operators.

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Change memory format promotion rules of point wise operators."




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Change memory format promotion rules of point wise operators."




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

VitalyFedyunin commented:
I wonder it is better to add wiki url to the warning message with more detailed help on mixed memory format operations.

ngimel commented:
Yeah, adding wiki url makes sense, otherwise the warning won't tell much to people who are accidentally getting into this area, not trying to use channels-last.

VitalyFedyunin commented:
Ping @gchanan @ngimel

ngimel commented:
This PR looks fine for what it tries to do, however, given that it now favors layout of the first tensor, how different is it from former TI behavior without any special-casing? It would also favor the first tensor layout, but the code would be much simpler. There's a couple of differing behaviors: adding channels-last NC11 to a full-size 4d contiguous tensor - with former TI behavior it would result in a contiguous tensor, now it is still channels last. Similarly, unary op on NC11 channels last formerly would result in tensor with contiguous strides (which is compatible with numpy), but now (even before this PR) it preserves the cl strides. 
My problem is that code is now too complicated, and propagation rules are too complicated. We should think about how to make them simpler. If making them simpler is impossible, we should document them, including in the code - currently it is very hard to understand what's going on.

VitalyFedyunin commented:
I will put promotions wiki rules tomorrow and link here and inside the warning message.

VitalyFedyunin committed:
Rebase to master on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

VitalyFedyunin committed:
Add propagation rules test
 on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

ngimel commented on test/test_torch.py:
```diff
@@ -14109,6 +14109,27 @@ def test_helper(shape, numel, memory_format, device):
+    def test_memory_format_proparation_rules(self, device):
+        contiguous = torch.rand(10, 3, 5, 5, device=device) 
+        cl = torch.rand(10, 3, 5, 5, device=device).contiguous(memory_format=torch.channels_last)
+        ambiguous = torch.rand(10, 3, 1, 1, device=device).contiguous(memory_format=torch.channels_last)
```
I'm afraid this test isn't testing what it thinks it is testing. `.contiguous(memory_format)` is a light call that does not force-reset the strides, what you need here is `.to(memory_format)`:
```
In [4]: ambiguous = torch.rand(10, 3, 1, 1, device="cuda").contiguous(memory_format=torch.channels_last)                                                                                                                

In [5]: ambiguous.stride()                                                                                                                                                                                              
Out[5]: (3, 1, 1, 1)

In [6]: ambiguous = torch.rand(10, 3, 1, 1, device="cuda").to(memory_format=torch.channels_last)                                                                                                                        

In [7]: ambiguous.stride()                                                                                                                                                                                              
Out[7]: (3, 1, 3, 3)
```
For the latter case (even though it's still ambiguous) `.suggest_memory_format` will return `channels_last`, for the former it will return `contiguous`.

ngimel replied:
I verified that the test fails if I replace `.contiguous` here with `.to`

VitalyFedyunin replied:
Thanks! Fixed.

VitalyFedyunin committed:
Fix .to cases on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

VitalyFedyunin committed:
Fix .to cases on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

VitalyFedyunin committed:
Fix formatting on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last




[ghstack-poisoned]

ngimel commented:
This now loses channels-last on ambiguous channels-last + bias, or op(channels-last, scalar), which can be a relatively common occurrence in GANs
```
In [15]: ambiguous = torch.rand(10, 3, 1, 1, device=device).to(memory_format=torch.channels_last)                                                                                                                       

In [16]: ambiguous.stride()                                                                                                                                                                                             
Out[16]: (3, 1, 3, 3)

In [17]: bias=torch.randn(3,1,1, device=device)                                                                                                                                                                         

In [18]: (ambiguous+bias).stride()                                                                                                                                                                                      
Out[18]: (3, 1, 1, 1)

In [19]: (ambiguous*5).stride()                                                                                                                                                                                         
Out[19]: (3, 1, 1, 1)
```

VitalyFedyunin commented:
If I remember correctly we decided to ignore (document) ambiguity and loss of channels last in GAN until we implement permutations.

jjsjann123 commented:
> This now loses channels-last on ambiguous channels-last + bias, or op(channels-last, scalar), which can be a relatively common occurrence in GANs
> 
> ```
> In [15]: ambiguous = torch.rand(10, 3, 1, 1, device=device).to(memory_format=torch.channels_last)                                                                                                                       
> 
> In [16]: ambiguous.stride()                                                                                                                                                                                             
> Out[16]: (3, 1, 3, 3)
> 
> In [17]: bias=torch.randn(3,1,1, device=device)                                                                                                                                                                         
> 
> In [18]: (ambiguous+bias).stride()                                                                                                                                                                                      
> Out[18]: (3, 1, 1, 1)
> 
> In [19]: (ambiguous*5).stride()                                                                                                                                                                                         
> Out[19]: (3, 1, 1, 1)
> ```
I agree with Natalia's concern here.

Per our documentation, pointwise operators preserves memory format.

Is there a good reason that we don't do that for ambiguous cases here? Because putting the behavior there is intentionally introducing discrepancy between documentation and implementation.
Given that we also encourage user to investigate operations that does not propagate memory format properly, this is going to get us many opened issues. We really need a strong reason to keep the implementation as-is.

VitalyFedyunin commented:
There are no implementation concerns, I was trying to avoid the situation when our rules include a lot of corner cases. I'm going to change it.

ngimel commented:
I'd say if there is an unambiguous operand, it should take precedence, but if all operands are ambiguous (as is the case with adding bias and ops with scalars), then if one of them identifies as channels last it should be propagated.

VitalyFedyunin committed:
Fix ambigous + bias case on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last


Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix ambigous + bias case on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last


Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

[ghstack-poisoned]

VitalyFedyunin commented:
@ngimel ready for review

ngimel reviewed (approved):
Looks functionally correct now, I hope it does not affect perf too much.

VitalyFedyunin commented:
@ailzhang `test_memory_format_proparation_rules_xla` strangely fails here, looks like `xla` works differently with broadcasting

VitalyFedyunin committed:
Fix perf on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last


Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix perf on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last


Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -972,6 +1024,8 @@ bool TensorIterator::fast_set_up() {
             //       We will revisit this logic when doing the TensorIterator refactor work and
             //       it would probably be better to move this logic out of TensorIterator.
 
+            // TODO(vitalyf): Potentionally might try to address bigger than original memory size
```
making size sufficient is taken care of formerly in compute_shape and now in resize_outputs, so this is fine.

gchanan commented:
We decided to move this to 1.6.0.

VitalyFedyunin committed:
Merge to master on "Change memory format promotion rules of point wise operators."

Modify memory format promotion rules to avoid promoting when one of the input is ambiguous. New rules are:
 Ambiguous + Contiguous = Contiguous
 Ambiguous + Channels Last = Channels Last
 Contiguous + Ambiguous ( NC11 ) = Contiguous
 Contiguous + Channels Last = Contiguous ( + Warning )  Before this PR: Channels Last
 Channels Last + Contiguous = Channels Last ( + Warning )
 Channels Last + Ambiguous = Channels Last
 Bias + Channels Last = Channels Last
 Channels Last + Bias = Channels Last


Differential Revision: [D21819573](https://our.internmc.facebook.com/intern/diff/D21819573)

[ghstack-poisoned]
